File: griffon-app/controllers/ru/dkolmogortsev/TaskPanelController.java
package ru.dkolmogortsev;

import com.google.common.collect.Lists;
import griffon.core.artifact.GriffonController;
import griffon.inject.MVCMember;
import griffon.metadata.ArtifactProviderFor;
import java.util.List;
import java.util.Map;
import javafx.collections.ObservableMap;
import javax.annotation.Nonnull;
import javax.inject.Inject;
import org.codehaus.griffon.runtime.core.artifact.AbstractGriffonController;
import org.joda.time.LocalDate;
import ru.dkolmogortsev.messages.TaskStopped;
import ru.dkolmogortsev.messages.TimeEntryDeleted;
import ru.dkolmogortsev.task.TimeEntry;
import ru.dkolmogortsev.task.storage.TimeEntriesStorage;

/**
 * Created by dkolmogortsev on 2/7/17.
 */
@ArtifactProviderFor(GriffonController.class)
public class TaskPanelController extends AbstractGriffonController
{

    @MVCMember
    private TaskPanelModel model;

    @Inject
    private TimeEntriesStorage timeEntriesStorage;

    public void onTaskStopped(TaskStopped taskStopped)
    {
        TimeEntry timeEntry = timeEntriesStorage.get(taskStopped.getId());

        model.getMap().compute(timeEntry.getEntryDate(), (s, timeEntries) -> timeEntriesStorage.getByEntryDate(s));
    }

    public void onTimeEntryDeleted(TimeEntryDeleted deleted)
    {
        TimeEntry timeEntry = timeEntriesStorage.delete(deleted.getEntryId());
        model.getMap().compute(timeEntry.getEntryDate(), (s, timeEntries) ->
        {
            List<TimeEntry> es = timeEntriesStorage.getByEntryDate(s);
            return es;
        });
    }

    @Override
    public void mvcGroupInit(
            @Nonnull
                    Map<String, Object> args)
    {
        super.mvcGroupInit(args);
        initData();
    }

    public void initData()
    {
        ObservableMap<Long, List<TimeEntry>> modelMap = model.getMap();
        timeEntriesStorage.getEntriesGroupedByDay().entrySet().stream().forEachOrdered(stringListEntry ->
        {
            modelMap.put(stringListEntry.getKey(), stringListEntry.getValue());
        });
        if (modelMap.isEmpty())
        {
            modelMap.put(LocalDate.now().toDate().getTime(), Lists.newArrayList());
        }
    }

}

File: griffon-app/controllers/ru/dkolmogortsev/TaskPanelController.kt
package ru.dkolmogortsev

import com.google.common.collect.Lists
import griffon.core.artifact.GriffonController
import griffon.inject.MVCMember
import griffon.metadata.ArtifactProviderFor
import org.codehaus.griffon.runtime.core.artifact.AbstractGriffonController
import org.joda.time.LocalDate
import ru.dkolmogortsev.events.TaskStoppedEvent
import ru.dkolmogortsev.events.TimeEntryDeletedEvent
import ru.dkolmogortsev.task.TimeEntry
import ru.dkolmogortsev.task.storage.TimeEntriesStorage
import javax.inject.Inject

/**
 * Created by dkolmogortsev on 2/7/17.
 */
@ArtifactProviderFor(GriffonController::class)
class TaskPanelController : AbstractGriffonController()
{
    @MVCMember
    lateinit var model: TaskPanelModel
    @Inject
    lateinit var timeEntriesStorage: TimeEntriesStorage

    fun onTaskStopped(taskStoppedEvent: TaskStoppedEvent)
    {
        val timeEntry = timeEntriesStorage.get(taskStoppedEvent.id)

        (model.map as java.util.Map<Long, List<TimeEntry>>).compute(timeEntry.entryDate) { s, timeEntries -> timeEntriesStorage.getByEntryDate(s) }
    }

    fun onTimeEntryDeleted(deletedEvent: TimeEntryDeletedEvent)
    {
        val timeEntry = timeEntriesStorage.delete(deletedEvent.entryId)
        (model.map as java.util.Map<Long, List<TimeEntry>>).compute(timeEntry.entryDate) { s, timeEntries ->
            val es = timeEntriesStorage.getByEntryDate(s)
            es
        }
    }

    override fun mvcGroupInit(
            args: Map<String, Any>)
    {
        super.mvcGroupInit(args)
        initData()
        val eventRouter = getApplication().eventRouter
        eventRouter.addEventListener(TaskStoppedEvent::class.java, { onTaskStopped(it!!.first() as TaskStoppedEvent) })
        eventRouter.addEventListener(TimeEntryDeletedEvent::class.java, { onTimeEntryDeleted(it!!.first() as TimeEntryDeletedEvent) })
    }

    fun initData()
    {
        val modelMap = model.map
        timeEntriesStorage.entriesGroupedByDay.entries.stream()
                .forEach { stringListEntry -> modelMap.put(stringListEntry.key, stringListEntry.value) }
        if (modelMap.isEmpty())
        {
            modelMap.put(LocalDate.now().toDate().time, Lists.newArrayList<TimeEntry>())
        }
    }
}
